## 1. Import dataset

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
import sklearn
from sklearn.model_selection import train_test_split


In [2]:
data = pd.read_csv("findata_challenge_train.csv")

In [3]:
data.head()

x0        x1        x2        x3        x4        x5   x6   x7  \
0  0.067563 -2.172492 -1.713774 -0.146854 -0.685219 -0.933838  4.0  2.0   
1  0.099622  0.584803 -6.089758 -0.252186 -0.334912 -4.671281  4.0  2.0   
2  0.648421  3.881137  0.961359  0.217361 -1.471850 -0.975352  5.0  1.0   
3  0.005775 -2.726153  2.891378 -0.038325  2.280847  1.761478  3.0  3.0   
4  0.044396  2.803576 -1.923381 -0.116657 -1.574199  6.098627  3.0  1.0   

          x8         x9  ...       x116  x117  x118      x119      x120  \
0  -4.476051  -1.472306  ...   4.607926  52.0   1.0  0.773809  0.766531   
1  -0.387076  -0.664552  ...  15.889763  53.0   1.0  7.065852  0.994304   
2  -0.850146  -4.011487  ...  13.770410  53.0   1.0 -8.598553 -2.958302   
3   9.116188  -4.610151  ...   5.904901  50.0   1.0  8.194975 -7.798308   
4 -13.956141  13.785099  ... -14.603703  47.0   0.0  2.330789 -8.249825   

        x121       x122       x123      x124  y  
0   6.307142   9.074515  -2.694049  0.330964  0  
1  -0.028411 -21.472850   5.626095  1.293455  0  
2 -12.784030 -36.855873   0.356093  1.499790  0  
3  10.910515  15.226094 -10.440813 -0.735116  0  
4 -12.680490  40.192302  -5.037065  1.679262  1  

[5 rows x 126 columns]

## 2. Exploratory Data Analysis (EDA)

In [4]:
print('Ukuran data = {}'.format(data.shape))
print('Index column data = {}'.format(data.columns))

Ukuran data = (100000, 126)
Index column data = Index(['x0', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9',
       ...
       'x116', 'x117', 'x118', 'x119', 'x120', 'x121', 'x122', 'x123', 'x124',
       'y'],
      dtype='object', length=126)


data set terdiri dari 125 features dan 1 kolom target dengan nilai 1 (churn) atau 0 (tidak churn)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 126 entries, x0 to y
dtypes: float64(120), int64(1), object(5)
memory usage: 96.1+ MB


dari info data diatas, diketahui ada 5 features yang memiliki tipe data object sehingga perlu dicek lebih lanjut tipe data di 5 kolom ini.

In [6]:
data.describe()

x0            x1            x2            x3            x4  \
count  99877.000000  99869.000000  99876.000000  99885.000000  99863.000000   
mean      -0.000110      0.001045      0.009028     -0.000254     -0.677076   
std        0.424300      3.379969      3.088269      0.121542      2.561589   
min       -1.724571    -16.097070    -13.519334     -0.517647    -13.627396   
25%       -0.288149     -2.274297     -2.070010     -0.082583     -2.384499   
50%       -0.000976      0.015052      0.014444     -0.000445     -0.679873   
75%        0.286196      2.288710      2.087186      0.081461      1.034269   
max        1.762542     14.768077     13.335437      0.513763     10.442234   

                 x5            x6            x7            x8            x9  \
count  99884.000000  99878.000000  99867.000000  99866.000000  99870.000000   
mean      -0.002404      4.306714      2.366517      0.013684     -0.016469   
std        4.042530      1.140149      0.728027      7.766165      5.391990   
min      -16.984482      0.000000      0.000000    -34.513401    -23.449782   
25%       -2.732918      4.000000      2.000000     -5.233073     -3.642062   
50%        0.000063      4.000000      2.000000      0.004569     -0.034581   
75%        2.726788      5.000000      3.000000      5.273118      3.617453   
max       20.186745      9.000000      5.000000     33.000825     23.296840   

       ...          x116          x117          x118          x119  \
count  ...  99877.000000  99864.000000  99873.000000  99878.000000   
mean   ...      0.038825     50.012697      0.577373      3.883003   
std    ...      9.063181      6.892136      0.493980     10.795317   
min    ...    -37.320446     21.000000      0.000000    -42.113882   
25%    ...     -6.077436     45.000000      0.000000     -3.247750   
50%    ...      0.044680     50.000000      1.000000      3.900552   
75%    ...      6.144167     54.000000      1.000000     11.039607   
max    ...     38.712294     80.000000      1.000000     57.003256   

               x120          x121          x122          x123          x124  \
count  99880.000000  99871.000000  99873.000000  99870.000000  99877.000000   
mean      -0.021634     -0.015191      7.594731     -0.012175      0.008854   
std        5.571357      9.745984     35.358657      3.947540      4.128462   
min      -23.512064    -43.961674   -165.068134    -17.229400    -16.394975   
25%       -3.776839     -6.630953    -16.275026     -2.656104     -2.766022   
50%       -0.018491     -0.025400      6.668708     -0.032674     -0.002028   
75%        3.743165      6.594418     30.879552      2.638333      2.788782   
max       25.927916     41.261627    174.645087     17.058603     17.921770   

                   y  
count  100000.000000  
mean        0.169920  
std         0.375564  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 121 columns]

## 3. Feature Engineering

pada dataset terdapat 5 kolom yang memiliki tipe data kategorikal dan data berupa object/string. oleh karena itu data ini harus diubah menjadi nilai integer supaya dapat dilakukan proses training machine learning

In [7]:
data[["x21", "x79", "x89",  "x108", "x112"]].head()

x21        x79               x89     x108 x112
0  Jun  Laki-laki             Jambi      SMP  May
1  Jul  Perempuan    Sumatera Barat      SMP  May
2  Jul  Perempuan  Sumatera Selatan       SD  Jun
3  Jun  Laki-laki  Kalimantan Utara      SMP  Sep
4  Jul  Laki-laki  Kalimantan Utara  SMA/SMK  Jun

sebelum dilakukan konversi data object menjadi nilai integer 0,1,2 ..., perlu diketahui komponen data pada setiap kolom seperti dibawah ini.

In [8]:
data["x21"].value_counts().index

Index(['Jun', 'Jul', 'May', 'Aug', 'Apr', 'Sep', 'Mar', 'Oct', 'Nov', 'Feb',
       'Jan', 'Dec'],
      dtype='object')

In [9]:
data["x79"].value_counts().index

Index(['Perempuan', 'Laki-laki'], dtype='object')

In [10]:
data["x89"].value_counts().index

Index(['Kalimantan Tengah', 'Kalimantan Barat', 'Kalimantan Timur',
       'Sumatera Utara', 'Kalimantan Selatan', 'Sumatera Selatan',
       'Kalimantan Utara', 'Sumatera Barat', 'Gorontalo', 'Jambi',
       'Sulawesi Barat', 'Bengkulu', 'Sulawesi Tengah', 'Bangka Belitung',
       'Sulawesi Tenggara', 'Jawa Timur', 'Sulawesi Utara', 'Jawa Tengah',
       'Bali', 'Jawa Barat', 'NTB', 'DI Yogyakarta', 'NTT', 'Banten', 'Maluku',
       'Maluku Utara', 'DKI Jakarta'],
      dtype='object')

In [11]:
data["x108"].value_counts().index

Index(['SMP', 'SMA/SMK', 'SD', 'S1/S2/S3', 'Tidak sekolah'], dtype='object')

In [12]:
data["x112"].value_counts().index

Index(['Jul', 'Jun', 'Aug', 'May', 'Sep', 'Apr', 'Oct', 'Mar', 'Nov', 'Feb',
       'Dec', 'Jan'],
      dtype='object')

data kategorikal akan dilakukan proses encoding untuk mengubah data object menjadi tipedata integer.

In [13]:
encoding_info = {
    "x21":{
           "Jan" : 1, 
           "Feb" : 2, 
           "Mar" : 3, 
           "Apr" : 4, 
           "May" : 5, 
           "Jun" : 6, 
           "Jul" : 7, 
           "Aug" : 8, 
           "Sep" : 9, 
           "Oct" : 10, 
           "Nov" : 11, 
           "Dec" : 12, 
    },
    
    "x79":{
           "Perempuan" : 0, 
           "Laki-laki" : 1,
    },
    
    "x89":{
           "Kalimantan Tengah" : 1, 
           "Kalimantan Barat" : 2, 
           "Kalimantan Timur" : 3,
           "Sumatera Utara" : 4, 
           "Kalimantan Selatan" : 5, 
           "Sumatera Selatan" : 6,
           "Kalimantan Utara" : 7, 
           "Sumatera Barat" : 8, 
           "Gorontalo" : 9, 
           "Jambi" : 10,
           "Sulawesi Barat" : 11, 
           "Bengkulu" : 12, 
           "Sulawesi Tengah" : 13, 
           "Bangka Belitung" : 14,
           "Sulawesi Tenggara" : 15, 
           "Jawa Timur" : 16, 
           "Sulawesi Utara" : 17, 
           "Jawa Tengah" : 18,
           "Bali" : 19, 
           "Jawa Barat" : 20, 
           "NTB" : 21, 
           "DI Yogyakarta" : 22, 
           "NTT" : 23, 
           "Banten" : 24, 
           "Maluku" : 25,
           "Maluku Utara" : 26, 
           "DKI Jakarta" : 27,
    },
    
    "x108":{
            "SMP" : 1, 
            "SMA/SMK" : 2, 
            "SD" : 3, 
            "S1/S2/S3" : 4, 
            "Tidak sekolah" : 5,
    },
    
    "x112":{
           "Jan" : 1, 
           "Feb" : 2, 
           "Mar" : 3, 
           "Apr" : 4, 
           "May" : 5, 
           "Jun" : 6, 
           "Jul" : 7, 
           "Aug" : 8, 
           "Sep" : 9, 
           "Oct" : 10, 
           "Nov" : 11, 
           "Dec" : 12,
    },
}

In [14]:
data = data.replace(encoding_info)

setelah dilakukan proses encoding, didapat 126 kolom dengan 125 tipedata float dan 1 kolom dengan tipe data integer

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 126 entries, x0 to y
dtypes: float64(125), int64(1)
memory usage: 96.1 MB


In [16]:
#null data
data = data.dropna()

## 4. Modeling

In [17]:
f"ukuran data = {data.shape}"

'ukuran data = (83976, 126)'

pada data train dilakukan pembagian data test dan train. dimana data test akan digunakan sebagai pengecekan performansi diakhir training. ukuran data test sebesar 20 % dari total data train

In [18]:
train_data, test_data = train_test_split(data, test_size=0.2, random_state=1)

In [19]:
train_data.head()

x0        x1        x2        x3        x4        x5   x6   x7  \
50136  0.716442 -1.238224 -4.359058 -0.203935 -4.195777  2.614011  5.0  3.0   
25404  0.439535 -3.646541 -1.212632  0.004336 -3.414677  5.627984  3.0  3.0   
88792  1.085459 -2.537893 -3.019716 -0.160093  0.254003 -5.720166  6.0  3.0   
84312 -0.409694  1.402932 -0.450879 -0.045943 -2.184713 -6.337178  4.0  2.0   
34871 -0.226736 -6.910791  1.288814 -0.064964 -2.562597 -3.545132  5.0  2.0   

              x8         x9  ...       x116  x117  x118       x119      x120  \
50136   1.689884   3.501862  ...  19.434997  62.0   0.0   4.971541  0.956939   
25404  -6.175962   2.646168  ...   2.833920  54.0   0.0  16.316122 -0.021450   
88792  25.297958 -14.391879  ... -11.862977  52.0   1.0  -1.855619  5.445247   
84312 -19.182505   1.380707  ...  -9.702033  62.0   1.0 -20.920315 -2.419671   
34871   8.725936  -4.954979  ...  -2.545722  49.0   0.0  -2.859286 -5.358387   

            x121       x122      x123      x124  y  
50136  -5.720115  -7.737655  3.117346 -3.539042  0  
25404  15.257600  17.410439  1.312798  3.215625  0  
88792 -10.613543  14.520623 -3.634796 -0.919308  0  
84312  -1.502155 -30.560463  2.608436  2.188769  0  
34871  -3.443853  30.912245  1.886077 -6.584996  0  

[5 rows x 126 columns]

In [20]:
X_train = train_data.drop(["y"], axis = 1)

In [21]:
y_train = train_data[["y"]]

In [22]:
X_test = test_data.drop(["y"], axis = 1)

In [23]:
y_test = test_data[["y"]]

In [24]:
len(X_train.iloc[0])

125

## 5. Create model using Neural Network

model machine learning menggunakan model Neural Network. Jumlah layer yang digunakan sebanyak 8 layer dimana terdiri dari 1 input laye, 6 hidden layer dan 1 output layer. 1 input layer berupa flatten layer yang memastikan input data memiliki dimensi / jumlah data sebanyak 125, hidden layer memiliki 2 jenis activation function yakni ELU dan RELU kombinasi ini dilakukan untuk meningkatkan performansi secara quality dan kecepatan. adapun dilakukan kombinasi dengan memberikan dropout/membuang beberapa neuron dengan probability 20 % untuk meningkatan performansi model.

In [25]:
model5 = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(125,)),
    tf.keras.layers.Dense(500, activation='elu'),
    tf.keras.layers.Dense(500, activation='relu'),
    tf.keras.layers.Dense(500, activation='elu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(500, activation='relu'),
    tf.keras.layers.Dense(500, activation='elu'),
    tf.keras.layers.Dense(500, activation='relu'),
    tf.keras.layers.Dense(1, activation = 'sigmoid')
])

In [26]:
model5.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 125)               0         
                                                                 
 dense (Dense)               (None, 500)               63000     
                                                                 
 dense_1 (Dense)             (None, 500)               250500    
                                                                 
 dense_2 (Dense)             (None, 500)               250500    
                                                                 
 dropout (Dropout)           (None, 500)               0         
                                                                 
 dense_3 (Dense)             (None, 500)               250500    
                                                                 
 dense_4 (Dense)             (None, 500)               2

In [27]:
# optimizer = tf.keras.optimizers.SGD(learning_rate = 0.01)

model5.compile(optimizer = "adam",
             loss = 'binary_crossentropy',
             metrics=['accuracy'])

training dilakukan dengan jumlah epoch atau training sebanyak 10 kali untuk seluruh data yang ada

In [28]:
model5.fit(X_train, y_train, epochs=20)

Epoch 1/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.5716 - accuracy: 0.8397
Epoch 2/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3771 - accuracy: 0.8681
Epoch 3/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3472 - accuracy: 0.8847
Epoch 4/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3405 - accuracy: 0.8889
Epoch 5/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3314 - accuracy: 0.8936
Epoch 6/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3334 - accuracy: 0.8939
Epoch 7/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3328 - accuracy: 0.8940
Epoch 8/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3223 - accuracy: 0.8981
Epoch 9/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.3248 - accuracy: 0.8977
Epoch 10/20
2100/2100 [==============================] - 8s 4ms/step - loss: 0.317

## 6. Evaluation

untuk data test sebanyak 20% dari data training, model memiliki performansi akurasi sebesar 90,6 %

In [29]:
model5.evaluate(X_test, y_test)

525/525 [==============================] - 1s 1ms/step - loss: 0.3104 - accuracy: 0.9038


[0.3103559613227844, 0.9037865996360779]

In [30]:
y_prob = model5.predict(X_test)
y_prob

array([[0.10053828],
       [0.09777778],
       [0.9140174 ],
       ...,
       [0.09681064],
       [0.09872174],
       [0.10001844]], dtype=float32)

In [31]:
y_pred = y_prob.round()
print(y_pred)

[[0.]
 [0.]
 [1.]
 ...
 [0.]
 [0.]
 [0.]]


## 7. Prediction Output

input data test challenge pada suatu variable

In [32]:
findata_challenge_test = pd.read_csv("findata_challenge_test.csv")

In [33]:
findata_challenge_test

x0        x1        x2        x3        x4        x5   x6   x7  \
0      0.403735 -1.966104 -1.322339  0.084642  2.481997 -1.719155  5.0  2.0   
1      0.032128  6.516294 -0.939752  0.026937 -0.992999  5.995462  4.0  3.0   
2     -0.280345  0.930291 -1.028828 -0.050924 -3.164733  6.885108  3.0  2.0   
3     -0.313420 -0.568857 -1.576051 -0.003607  5.539969 -1.594770  4.0  3.0   
4      0.108825 -4.815631  5.491084 -0.123928 -1.932024  4.014821  5.0  3.0   
...         ...       ...       ...       ...       ...       ...  ...  ...   
24995 -0.177365 -1.441883  1.644567  0.135914 -1.434104 -2.197283  6.0  2.0   
24996  0.219625  2.084640  3.710896  0.035350  0.792121  3.417730  5.0  2.0   
24997 -0.204578 -0.332965 -1.205238  0.202758 -0.147738 -0.664850  4.0  3.0   
24998 -0.298675  0.750118 -2.856962  0.167308  0.056763 -5.628873  4.0  3.0   
24999 -0.092210  0.602218  1.042477  0.146612 -1.293322 -2.077064  3.0  2.0   

              x8        x9  ...      x115       x116  x117  x118       x119  \
0     -11.546670 -7.073770  ...  3.580278  -7.765934  59.0   1.0  -9.897690   
1      -5.634876  1.937998  ...  5.783655  15.667015  47.0   1.0   7.781887   
2       1.609005 -1.812488  ... -0.389116   8.523648  55.0   0.0   6.815214   
3     -11.517734 -5.364282  ...  3.930974 -13.912115  41.0   0.0   3.158137   
4      -3.117576  2.004889  ... -4.056091 -14.432712  52.0   0.0  -3.867657   
...          ...       ...  ...       ...        ...   ...   ...        ...   
24995 -12.595754  1.800405  ... -2.589491 -27.093773  57.0   1.0 -16.034945   
24996   4.568492 -6.418023  ...  2.248717  -0.310714  58.0   1.0  -0.214017   
24997  -5.961850 -6.291984  ...  5.194106   6.376269  45.0   0.0  14.348007   
24998  -0.477643  4.446171  ... -1.079791  10.990180  41.0   1.0  11.317455   
24999  -6.101376  0.354084  ...  2.611718   1.372462  38.0   0.0   7.830794   

            x120       x121       x122      x123       x124  
0       5.694125  -6.734618  17.052251  3.601040   2.873498  
1       9.752716   5.918369  15.922908  5.460557  10.475816  
2      -3.325532   4.123546 -19.656038  8.380318  -1.203315  
3      -7.088875  -3.035983  34.277048 -3.169203   1.417681  
4     -13.308519   1.008302  41.449381 -0.804560  -0.649453  
...          ...        ...        ...       ...        ...  
24995  -4.188319   1.857260 -47.994260 -1.111017   0.780211  
24996   2.988096   5.508766 -12.374789  2.773387  -6.644809  
24997   1.003478  -1.820065 -67.534113  1.520819   3.599916  
24998   6.262580  13.584108  29.077122 -0.135391  -2.858872  
24999   5.257363   0.975914  66.850592  2.770043  -0.984973  

[25000 rows x 125 columns]

In [34]:
findata_challenge_test = findata_challenge_test.replace(encoding_info)

isi kolom yang kosong dengan nilai rata-rata pada feature yang bersangkutan

In [35]:
findata_challenge_test = findata_challenge_test.fillna(findata_challenge_test.mean())

Hasil prediksi dalam probability

In [36]:
y_challenge = model5.predict(findata_challenge_test)
y_challenge

array([[0.09199679],
       [0.09800801],
       [0.09649867],
       ...,
       [0.92087597],
       [0.09770149],
       [0.09830537]], dtype=float32)

Hasil prediksi churn (1 :iya,0 :tidak)

In [37]:
y_challenge_class = y_challenge.round()
print(y_challenge_class)

[[0.]
 [0.]
 [0.]
 ...
 [1.]
 [0.]
 [0.]]


In [38]:
y_answer = []
for i in y_challenge_class:
    for x in i:
        y_answer.append(x)

In [39]:
y_answer = pd.DataFrame(y_answer, columns=['Predicted']).astype('int')

In [40]:
range_1 = list(range(0, 25000, 1))
Id = list(range_1)


In [41]:
y_challenge_table_fix = pd.DataFrame({"Id" : Id})

In [42]:
y_challenge_table_fix["Predicted"]  = y_answer['Predicted']

In [43]:
y_challenge_table_fix

Id  Predicted
0          0          0
1          1          0
2          2          0
3          3          0
4          4          0
...      ...        ...
24995  24995          0
24996  24996          0
24997  24997          1
24998  24998          0
24999  24999          0

[25000 rows x 2 columns]

Input/save data to csv file

In [44]:
y_challenge_table_fix.to_csv('kriesnawan_handy_sugianto-submission.csv', index = False)